params/json.libevm_test.go

```diff
@@ -127,20 +128,110 @@ func TestChainConfigJSONRoundTrip(t *testing.T) {
 			t.Cleanup(TestOnlyClearRegisteredExtras)
 			tt.register()
 
-			t.Run("json.Unmarshal()", func(t *testing.T) {
-				got := new(ChainConfig)
-				require.NoError(t, json.Unmarshal([]byte(tt.jsonInput), got))
-				require.Equal(t, tt.want, got)
-			})
+			expectedEncoded := bytes.NewBuffer(nil)
+			err := json.Compact(expectedEncoded, []byte(tt.jsonInput))
+			require.NoError(t, err)
 
-			t.Run("json.Marshal()", func(t *testing.T) {
-				var want bytes.Buffer
-				require.NoError(t, json.Compact(&want, []byte(tt.jsonInput)), "json.Compact()")
+			encoded, err := json.Marshal(tt.want)
+			require.NoError(t, err, "encoding error")
+			require.Equal(t, expectedEncoded.String(), string(encoded))
 
-				got, err := json.Marshal(tt.want)
-				require.NoError(t, err, "json.Marshal()")
-				require.Equal(t, want.String(), string(got))
-			})
+			decoded := new(ChainConfig)
+			err = json.Unmarshal(encoded, decoded)
+			require.NoError(t, err, "decoding error")
+			require.Equal(t, tt.want, decoded)
+		})
+	}
+}
+
+func TestUnmarshalChainConfigJSON(t *testing.T) {
+	t.Parallel()
+
+	type testExtra struct {
+		Field string `json:"field"`
+	}
+
+	testCases := map[string]struct {
+		jsonData       string // string for convenience
+		extra          any
+		reuseJSONRoot  bool
+		expectedConfig ChainConfig
+		expectedExtra  any
+		errMessage     string
+	}{
+		"invalid_json": {
+			extra:         &testExtra{},
+			expectedExtra: &testExtra{},
+			errMessage:    "decoding root chain config: unexpected end of JSON input",
+		},
+		"nil_extra": {
```

Comment:
I think we should keep it to `nil_extra` now that it must be a typed nil (so a bit less "empty" lol) since it's now back to a generic function accepting `extra *T` (it cannot be the untyped nil anymore)

Reply:
I'll keep it to `nil` because `empty` could mean `testExtra{}` (empty struct) if that makes some sense

I also changed the code to return an error `extra pointer argument is nil` because:

- I don't think there is a point using this function at all if the extra argument is a (typed) nil
- Previously, it was inconsistent behavior: it would error when re-using the JSON root (from `json.Unmarshal(data, extra)`) but would silently discard the "extra" JSON field without error if not re-using the JSON root.
- Discarding silently the extra field sounds like a dangerous path, better to return an error.